File: todo/app/src/main/java/com/k/todo/provider/MusicProvider.java
package com.k.todo.provider;


import com.k.todo.provider.data.MusicCache;

import java.lang.ref.WeakReference;
import java.util.ArrayList;

public class MusicProvider implements DataSourceListener {
    public static MusicProvider provider;

    public static MusicProvider create() {
        if (provider == null) {
            provider = new MusicProvider();
        }
        return provider;
    }

    private WeakReference<ArrayList<DataSourceListener>> listenerList;
    private MusicCache musicCache;

    public MusicProvider() {
        listenerList = new WeakReference<ArrayList<DataSourceListener>>(new ArrayList<DataSourceListener>());
        musicCache = MusicCache.create(this);
    }

    public void loadData(boolean reload) {
        musicCache.loadData(reload);
    }

    public void loadData() {
        musicCache.loadData(false);
    }

    public void registerDataSourceListener(DataSourceListener dataSourceListener) {
        ArrayList<DataSourceListener> listeners = listenerList.get();
        if (listeners == null || dataSourceListener == null) {
            return;
        }
        listeners.add(dataSourceListener);

    }

    public void unregisterDataSourceListener(DataSourceListener dataSourceListener) {
        ArrayList<DataSourceListener> listeners = listenerList.get();
        if (listeners == null || dataSourceListener == null) {
            return;
        }
        listeners.remove(dataSourceListener);

    }

    @Override
    public void onLoaded(Object datas) {
        ArrayList<DataSourceListener> listeners = listenerList.get();
        if (listeners == null) {
            return;
        }
        for (DataSourceListener listener : listeners) {
            listener.onLoaded(datas);
        }
    }

    @Override
    public void onLoadError(Object error) {
        ArrayList<DataSourceListener> listeners = listenerList.get();
        if (listeners == null) {
            return;
        }
        for (DataSourceListener listener : listeners) {
            listener.onLoadError(error);
        }
    }

    @Override
    public void onDataChanged(Object datas) {
        ArrayList<DataSourceListener> listeners = listenerList.get();
        if (listeners == null) {
            return;
        }
        for (DataSourceListener listener : listeners) {
            listener.onDataChanged(datas);
        }
    }

    @Override
    public void onLoading() {
        ArrayList<DataSourceListener> listeners = listenerList.get();
        if (listeners == null) {
            return;
        }
        for (DataSourceListener listener : listeners) {
            listener.onLoading();
        }
    }


}

File: todo/app/src/main/java/com/k/todo/provider/MusicProvider.kt
package com.k.todo.provider


import com.k.todo.provider.data.MusicCache

import java.lang.ref.WeakReference
import java.util.ArrayList

class MusicProvider : DataSourceListener {

    private val listenerList: ArrayList<DataSourceListener>?
    private val musicCache: MusicCache

    init {
        listenerList = ArrayList()
        musicCache = MusicCache.create(this)
    }

    fun loadData(reload: Boolean) {
        musicCache.loadData(reload)
    }

    fun loadData() {
        musicCache.loadData(false)
    }

    fun registerDataSourceListener(dataSourceListener: DataSourceListener?) {
        val listeners = listenerList
        if (listeners == null || dataSourceListener == null) {
            return
        }
        listeners.add(dataSourceListener)

    }

    fun unregisterDataSourceListener(dataSourceListener: DataSourceListener?) {
        val listeners = listenerList
        if (dataSourceListener == null) {
            return
        }
        listeners!!.remove(dataSourceListener)

    }

    override fun onLoaded(datas: Any) {
        val listeners = listenerList ?: return
        for (listener in listeners) {
            listener.onLoaded(datas)
        }
    }

    override fun onLoadError(error: Any) {
        val listeners = listenerList ?: return
        for (listener in listeners) {
            listener.onLoadError(error)
        }
    }

    override fun onDataChanged(datas: Any) {
        val listeners = listenerList ?: return
        for (listener in listeners) {
            listener.onDataChanged(datas)
        }
    }

    override fun onLoading() {
        val listeners = listenerList ?: return
        for (listener in listeners) {
            listener.onLoading()
        }
    }

    companion object {
        var provider: MusicProvider? = null

        fun create(): MusicProvider {
            if (provider == null) {
                provider = MusicProvider()
            }
            return provider!!
        }
    }


}
